rust/abacus-base/src/settings/chains.rs

```diff
@@ -109,142 +128,69 @@ impl<T> ChainSetup<T> {
-                        signer,
-                        Some(|| metrics.json_rpc_client_metrics()),
-                        Some((metrics.provider_metrics(), self.metrics_conf())),
-                    )
-                    .await?,
-            )),
-        }
+    ) -> eyre::Result<Box<dyn InterchainGasPaymaster>> {
+        self.try_into_contract(
+            signer,
+            metrics,
+            InterchainGasPaymasterBuilder {},
+            self.addresses.interchain_gas_paymaster.clone(),
+        )
+        .await
     }
 
-    /// Get a clone of the metrics conf with correctly configured contract
-    /// information.
-    pub fn metrics_conf(&self) -> PrometheusMiddlewareConf {
-        let mut cfg = self.metrics_conf.clone();
-
-        if cfg.chain.is_none() {
-            cfg.chain = Some(ChainInfo {
-                name: Some(self.name.clone()),
-            });
-        }
-
-        if let Ok(addr) = self.addresses.outbox.parse() {
-            cfg.contracts.entry(addr).or_insert_with(|| ContractInfo {
-                name: Some("outbox".into()),
-                functions: EthereumOutboxAbi::fn_map_owned(),
-            });
-        }
-        if let Some(igp) = &self.addresses.interchain_gas_paymaster {
-            if let Ok(addr) = igp.parse() {
-                cfg.contracts.entry(addr).or_insert_with(|| ContractInfo {
-                    name: Some("igp".into()),
-                    functions: EthereumInterchainGasPaymasterAbi::fn_map_owned(),
-                });
-            }
-        }
-        cfg
-    }
-}
-
-impl ChainSetup<InboxAddresses> {
-    /// Try to convert the chain setting into an inbox contract
-    pub async fn try_into_inbox(
+    /// Try to convert the chain setting into a Multisig Ism contract
+    pub async fn try_into_multisig_ism(
         &self,
         signer: Option<Signers>,
         metrics: &CoreMetrics,
-    ) -> eyre::Result<Box<dyn Inbox>> {
-        let metrics_conf = self.metrics_conf(metrics.agent_name(), &signer);
-        match &self.chain {
-            ChainConf::Ethereum(conf) => Ok(InboxBuilder {}
-                .make_with_connection(
-                    conf.clone(),
-                    &ContractLocator {
-                        chain_name: self.name.clone(),
-                        domain: self.domain.parse().expect("invalid uint"),
-                        address: self
-                            .addresses
-                            .inbox
-                            .parse::<ethers::types::Address>()?
-                            .into(),
-                    },
-                    signer,
-                    Some(|| metrics.json_rpc_client_metrics()),
-                    Some((metrics.provider_metrics(), metrics_conf)),
-                )
-                .await?),
-        }
+    ) -> eyre::Result<Box<dyn MultisigIsm>> {
+        self.try_into_contract(
+            signer,
+            metrics,
+            MultisigIsmBuilder {},
```

Comment:
A downside to this change is now we use the ethereum-specific `MultisigIsmBuilder` when we're not certain if the chain type is `Ethereum` yet. However I think we can change this in the future sort of easily by creating a chain-type-agnostic `MultisigIsmBuilder`. I'd leave that to a future PR though, just thinking out loud

Reply:
Ah indeed, good catch. Agreed on addressing this later.